openmeter/ent/schema/subscription_patch.go

```diff
@@ -0,0 +1,212 @@
+	"entgo.io/ent/schema"
+	"entgo.io/ent/schema/edge"
+	"entgo.io/ent/schema/field"
+	"entgo.io/ent/schema/index"
+
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type SubscriptionPatch struct {
+	ent.Schema
+}
+
+func (SubscriptionPatch) Annotations() []schema.Annotation {
+	return []schema.Annotation{
+		entsql.Annotation{Table: "subscription_patches"},
+	}
+}
+
+func (SubscriptionPatch) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionPatch) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_id").NotEmpty().Immutable(),
+		field.Time("applied_at").Immutable(),
+		field.Int("batch_index").Immutable(),
+		field.String("operation").NotEmpty().Immutable(),
+		field.String("path").NotEmpty().Immutable(),
+	}
+}
+
+func (SubscriptionPatch) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "subscription_id"),
+	}
+}
+
+func (SubscriptionPatch) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("subscription", Subscription.Type).Ref("subscription_patches").Field("subscription_id").Unique().Immutable().Required(),
+		edge.To("value_add_item", SubscriptionPatchValueAddItem.Type).Unique(),
+		edge.To("value_add_phase", SubscriptionPatchValueAddPhase.Type).Unique(),
+		edge.To("value_remove_phase", SubscriptionPatchValueRemovePhase.Type).Unique(),
+		edge.To("value_extend_phase", SubscriptionPatchValueExtendPhase.Type).Unique(),
+	}
+}
+
+// We create dedicated tables for each value type
+// This way data migration problems between different versions of patches mostly become
+// schema migration problems that we're staticly forced to address.
+
+type SubscriptionPatchValueAddItem struct {
+	ent.Schema
+}
+
+func (SubscriptionPatchValueAddItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+	}
+}
+
+func (SubscriptionPatchValueAddItem) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_patch_id").NotEmpty().Immutable(),
+		// Data properties
+		field.String("phase_key").NotEmpty().Immutable(),
+		field.String("item_key").NotEmpty().Immutable(),
+		field.String("feature_key").Optional().Nillable().Immutable(),
+		field.String("create_entitlement_entitlement_type").Optional().Nillable().Immutable(),
+		field.Float("create_entitlement_issue_after_reset").Optional().Nillable().Immutable(),
+		field.Uint8("create_entitlement_issue_after_reset_priority").Optional().Nillable().Immutable(),
+		field.Bool("create_entitlement_is_soft_limit").Optional().Nillable().Immutable(),
+		field.Bool("create_entitlement_preserve_overage_at_reset").Optional().Nillable().Immutable(),
+		field.String("create_entitlement_usage_period_iso_duration").Optional().Nillable().Immutable(),
+		field.JSON("create_entitlement_config", []byte{}).SchemaType(map[string]string{
```

Comment:
can we normalize this?

Reply:
This is the config of static entitlements, it cannot be normalized.